Initial Multiaddr Implementation

aria42 opened the pull request:
There are more protocols left to implement from [the partial spec](https://multiformats.io/multiaddr/), but thought this would be a good point to check in and get feedback, since the core pieces are here. 

I have the `Address` object support a collection view on the pairs of `[protocol-key value]`, which seems like the most natural way to use this data (e.g, configuring a network client). 

## Reading Guide

*  Start with `address.clj` to understand `Address` object and choice to have it support basic collection view of `[protocol-key value]` pairs. 
* `address/codec.clj` to understand `Codec`; the IP address implementation in particular was annoying since built-in platform libraries, specially on JS, almost have what you want, but require a decent amount of futzing. 

## Questions

* Some of the generic byte array manipulation seems generic and useful, should they be in `alphabase`?
* The `Codec` abstraction I'm using I think is needed since there's a few distinct formats for network protocols (e.g, DNS, IP, Port numbers, Onion, etc.). Should the interface be lower-level to use `offset` and `length` to allow direct manipulation of a buffer and avoid some byte array copies?

## TODO 
- [] Add UTF8 address codec for DNS protocol
- [] Add custom onion protocol 
- [] More tests, particularly on validation.

aria42 committed:
initial multiaddr protocol implementation/tests

greglook commented on src/multiformats/address.cljc:
```diff
@@ -2,29 +2,232 @@
+   (protocol-value-seq data 0))
+  ([^bytes data offset]
+   (when (< offset (alength data))
+       (lazy-seq
```
Indent is a bit off here, but I'll probably be running `cljfmt` over this anyway.

aria42 replied:
Ran `lein cljfmt fix` as its own commit.

greglook commented on src/multiformats/address.cljc:
```diff
@@ -2,29 +2,232 @@
+       (lazy-seq
+        (let [[code code-len] (varint/read-bytes data offset)
+              protocol-key (get code->protocol code)
+              {:keys [codec]} (ensure-protocol-attrs protocol-key)
```
Destructuring doesn't seem necessary here:
```suggestion
               codec (:codec (ensure-protocol-attrs protocol-key))
```

greglook commented on src/multiformats/address.cljc:
```diff
@@ -2,29 +2,232 @@
+        val-len-bytes (if (codec/fixed-byte-length codec)
+                        (b/byte-array 0)
+                        (varint/encode (alength ^bytes val-bytes)))]
+    (concat-arrs code-bytes val-len-bytes val-bytes)))
```
Feels like this could be more efficient by expanding the conditional tree, instead of allocating zero-length byte arrays. On the other hand, it's also not likely to be on any hot paths anytime soon.

greglook commented on src/multiformats/address.cljc:
```diff
@@ -2,29 +2,232 @@
+
+  #?(:cljs IEmptyableCollection)
+  (#?(:clj empty :cljs -empty)
+    [this] (Address. (b/byte-array 0) nil 0))
```
I believe that standard behavior is to preserve metadata on emptied collections:
```clojure
=> (-> {} (with-meta {:foo 123}) empty meta)
{:foo 123}
```

greglook commented on src/multiformats/address.cljc:
```diff
@@ -2,29 +2,232 @@
+  #?(:cljs IEquiv)
+  (#?(:clj equiv :cljs -equiv) [this other]
+    (and (instance? Address other)
+         (b/bytes= (.-_data this) (.-_data ^Address other))))
```
TIL Clojure can also use the hyphen-prefixed field access!

greglook commented on src/multiformats/address.cljc:
```diff
@@ -2,29 +2,232 @@
+(defn- parse-entries
+  [address-str]
+  (let [parts (->> #"/" (str/split address-str) (remove str/blank?))]
+    (loop [parts parts pairs []]
```
You could collapse the `let` above and build the parts list in this loop binding.

greglook commented on src/multiformats/address.cljc:
```diff
@@ -2,29 +2,232 @@
+              fixed-len (codec/fixed-byte-length codec)]
+          (if (and fixed-len (zero? fixed-len))
+            (recur (next parts) (conj pairs [protocol-key nil]))
+            (let [new-entry [protocol-key (str (second parts))]]
```
When is `(second parts)` not a string? I suppose this doesn't preclude an incomplete input like `/ip4`.

aria42 replied:
This might be unnecessary, but wanted to allow for integers (since some of these arguments are for ports and things that are naturally ints) and anything else that could be coerced to string this way, but happy to remove.

But I did update this to throw if we don't get a value for a protocol expecting one since I hadn't covered that.

greglook commented on src/multiformats/address.cljc:
```diff
@@ -2,29 +2,232 @@
+          (if (and fixed-len (zero? fixed-len))
+            (recur (next parts) (conj pairs [protocol-key nil]))
+            (let [new-entry [protocol-key (str (second parts))]]
+              (recur (drop 2 parts) (conj pairs new-entry)))))
```
Could use `(nnext parts)` for concision and eagerness.

greglook commented on src/multiformats/address.cljc:
```diff
@@ -2,29 +2,232 @@
+  (str (create [[:ip4 \"127.0.0.1\"][:tcp \"80\"]]))
+  >> \"/ip4/127.0.0.1/tcp/80\" "
+  [& entries]
+  (into (->Address (b/byte-array 0) nil 0) entries))
```
A more efficient approach here would be to directly encode the entries into a byte array before constructing the Address type. Without special reduction protocols, it seems like this is going to allocate O(n^2) memory.

aria42 replied:
Done, now this just calls `(map entry-bytes)` and then concatenates results so memory is now O(n)

greglook commented on src/multiformats/address/codec.cljc:
```diff
@@ -0,0 +1,165 @@
+      (b/byte-array 0))
+    (bytes->str [this data]
+      (when (pos? (alength ^bytes data))
+        (throw (ex-info "Expected 0 length array for no value protocol"
```
Ah, some of the zero-byte array stuff above makes more sense with these checks.

aria42 replied:
Ok, I actually left this in place defensively, but changed the call-site code to not bother creating 0 byte arrays

greglook commented on test/multiformats/address/codec_test.cljc:
```diff
@@ -0,0 +1,84 @@
+     (:import
+      clojure.lang.ExceptionInfo)))
+
+(defn ->bytes [bs]
```
🏌️ `b/init-bytes`.

greglook reviewed (commented):
Nice, thanks for tackling this! The spec is a little under-developed compared to the other multiformats, but it looks like you've got most of the useful pieces here. I left a number of comments and a couple of questions.

greglook commented:
> Some of the generic byte array manipulation seems generic and useful, should they be in alphabase?

Yeah, the slice-copy seems useful, as does concat-arrays. At this point alphabase is more useful as a byte-manipulation library than as a base-encoder. 😅

> The Codec abstraction I'm using I think is needed since there's a few distinct formats for network protocols (e.g, DNS, IP, Port numbers, Onion, etc.). Should the interface be lower-level to use offset and length to allow direct manipulation of a buffer and avoid some byte array copies?

A few of my comments were along these lines - since the codecs are an implementation detail, I think it'd be okay to ship this as an initial version and optimize more later. If you feel like tackling this though, I'm all for it.

aria42 committed:
cljfmt fix

aria42 committed:
Addressed PR comments

aria42 commented:
Ok, I addressed comments and changed just about everything you suggested, I'll make a separate PR in `alphabase` with versions of these functions, but leave these here for now and remove once that's merged.

aria42 committed:
use alpahabase/init-bytes for test

greglook commented on src/multiformats/address.cljc:
```diff
@@ -123,16 +119,19 @@
-    (concat-arrs code-bytes val-len-bytes val-bytes)))
+    (->> [code-bytes val-len-bytes val-bytes]
+         (remove nil?)
+         (apply concat-arrs))))
```
I think I'll update alphabase so that this function ignores nils, which would make this form simpler:
```clojure
(b/concat code-bytes val-len-bytes val-bytes)
```

greglook commented on src/multiformats/address.cljc:
```diff
@@ -123,16 +119,19 @@
 
 (deftype Address
-  [^bytes _data _meta ^:unsynchronized-mutable _hash]
+         [^bytes _data _meta ^:unsynchronized-mutable _hash]
```
Ah, I'm guessing this is the default cljfmt? This should only be indented one level.

aria42 replied:
Oh weird, sorry out of the box it treats `deftype` as another function instead of a `def`-style form.

greglook commented on src/multiformats/base.cljc:
```diff
@@ -7,16 +7,15 @@
   #?(:cljs
      (:require-macros
-       [multiformats.base :refer [defbase]]))
+      [multiformats.base :refer [defbase]]))
```
The other namespaces are already styled correctly, so I'd prefer you didn't reformat them.

greglook reviewed (approved):
Offset changes look good! I'll merge once you revert the changes to the other namespaces.

aria42 committed:
Address cljfmt, undo cljfmt to non multiaddr namespace

aria42 committed:
address concat-arrs nils (mostly retrigger circleci)

aria42 commented:
@greglook   I'm not sure why the CircleCI test is failing for cljs; I've got the latest `rhino` and `phantomjs` which appears to be what the test is using for dependencies. Locally, I'm running without error, but don't have a way to repro without getting access to CircleCI environment. 
<img width="569" alt="screen shot 2018-12-25 at 9 21 27 am" src="https://user-images.githubusercontent.com/31170/50425199-8387b800-0826-11e9-9bd8-fc242d23b176.png">

greglook commented:
Appears to be 5 instances of this error, three in the uint test and two in the utf8 test:
```
ERROR in (uint-codec-test) (at target/cljs/tests.js:NaN:4582)
round trip on numbers
expected: (= "80" (roundtrip codec/utf8-codec "80"))
  actual: #object[TypeError TypeError: second argument to Function.prototype.apply must be an array]
```

aria42 commented:
Oh I see the error in CircleCI, I just can't reproduce since tests passed on macOS with a clean build and the latest rhino; I'll add circleci to my fork so I can ssh into the build container and figure out what's going on. Alternatively, if you grant me admin on CircleCI I can do the same on your instance.

aria42 commented:
Ok so the issue is the version of `rhino` that CircleCI pulls in from Linux doesn't seem to have Typed Arrays and the underlying failure is that my UTF8 codec needs `Uint8Array` which doesn't exist on Rhino without some configuration that's non-trivial to pass in. I'm surprised this only broke on the recent push (was Rhino updated?) since all of my builds should've triggered this.
<img width="550" alt="screen shot 2018-12-25 at 1 33 33 pm" src="https://user-images.githubusercontent.com/31170/50426699-bdb68100-0849-11e9-8f2b-0d8a390eeef0.png">

If I replace the test to use `phantomjs` instead of `rhino` on the CircleCI server there's no problem since phantom seems to default to the language version that has typed arrays. 

So I can resolve this by changing the cljs test to use phantom environment instead. I'll put up a commit to do that and let you figure out how you want to deal.

aria42 committed:
Fix circleci test break (move to phantom + parametrize codec)

greglook commented:
🎉
